src/cargo/util/toml/mod.rs

```diff
@@ -2548,26 +2538,17 @@ pub fn validate_profile(
 /// This is a shallow check, which is reused for the profile itself and any overrides.
 fn validate_profile_layer(
     profile: &manifest::TomlProfile,
-    name: &str,
     cli_unstable: &CliUnstable,
     features: &Features,
 ) -> CargoResult<()> {
-    if let Some(codegen_backend) = &profile.codegen_backend {
+    if profile.codegen_backend.is_some() {
         match (
             features.require(Feature::codegen_backend()),
             cli_unstable.codegen_backend,
         ) {
             (Err(e), false) => return Err(e),
             _ => {}
         }
-
-        if codegen_backend.contains(|c: char| !c.is_ascii_alphanumeric() && c != '_') {
-            bail!(
-                "`profile.{}.codegen-backend` setting of `{}` is not a valid backend name.",
-                name,
-                codegen_backend,
-            );
-        }
```

Comment:
If preferred I can also keep this check on the cargo side when nightly features are disabled.

Reply:
How arbitrary you'd like to have for codegen-backend name? What is the real use case here?

The change seems fine to me though, as a nightly feature.

Reply:
Basically `-Zcodegen-backend` allows either of the following:

* The name of a builtin codegen backend. This will eventually be stabilized once one of the non-LLVM backends gets stabilized.
* The full path to a dylib implementing a codegen backend. **This will never be stable.** It exists to allow codegen backends not shipped as part of rustc to be used. For example cg_clif uses it when you locally build it or when using the version from its CI rather than the rustup distributed version (the CI version is newer). Same for cg_gcc when you locally build it.

Reply:
Thanks for the info! It makes sense to relax it a bit for now.

When stabilization, Cargo should only allow a set of known backends. We should track it somewhere, though we dont really have a Cargo tracking issue for codegen-backend right now.